packages/connect/src/ws_orderbook_channel.ts

```diff
@@ -1,166 +1,100 @@
-                        ),
-                        undefined,
-                    );
-                }
-                callback(undefined, this._connectionIfExists);
-            });
-            this._client.on(WebsocketClientEventType.ConnectFailed, error => {
-                callback(error, undefined);
-            });
-            this._client.connect(this._apiEndpointUrl);
+    private _handleWebSocketMessage(message: any): void {
+        if (_.isUndefined(message.data)) {
+            this._handler.onError(this, new Error(`Message does not contain utf8Data`));
+            return;
         }
-    }
-    private _handleWebSocketMessage(
-        requestId: number,
-        subscriptionOpts: OrderbookChannelSubscriptionOpts,
-        message: WebSocket.IMessage,
-        handler: OrderbookChannelHandler,
-    ): void {
-        if (!_.isUndefined(message.utf8Data)) {
-            try {
-                const utf8Data = message.utf8Data;
-                const parserResult = orderbookChannelMessageParser.parse(utf8Data);
-                if (parserResult.requestId === requestId) {
-                    switch (parserResult.type) {
-                        case OrderbookChannelMessageTypes.Snapshot: {
-                            handler.onSnapshot(this, subscriptionOpts, parserResult.payload);
-                            break;
-                        }
-                        case OrderbookChannelMessageTypes.Update: {
-                            handler.onUpdate(this, subscriptionOpts, parserResult.payload);
-                            break;
-                        }
-                        default: {
-                            handler.onError(
-                                this,
-                                subscriptionOpts,
-                                new Error(`Message has missing a type parameter: ${utf8Data}`),
-                            );
-                        }
-                    }
+        try {
+            const utf8Data = message.data;
+            const parserResult = orderbookChannelMessageParser.parse(utf8Data);
+            const subscriptionOpts = this._subscriptionOptsList[parserResult.requestId];
+            if (_.isUndefined(subscriptionOpts)) {
+                this._handler.onError(this, new Error(`Message has unknown requestId: ${utf8Data}`));
+                return;
+            }
+            switch (parserResult.type) {
+                case OrderbookChannelMessageTypes.Snapshot: {
+                    this._handler.onSnapshot(this, subscriptionOpts, parserResult.payload);
+                    break;
+                }
+                case OrderbookChannelMessageTypes.Update: {
+                    this._handler.onUpdate(this, subscriptionOpts, parserResult.payload);
+                    break;
+                }
+                default: {
+                    this._handler.onError(this, new Error(`Message has unknown type parameter: ${utf8Data}`));
```

Comment:
This feels nicer. One handler gets this error and maybe can do something about it. We do have an available subscriptionOpts at this point so we could be more explicit about the subscription which caused this message (which may allow a handler to take an action like re-subscribe). It would require an optional parameter in the function definition.

```
onError: (
         err: Error,
         subscriptionOpts?: OrderbookChannelSubscriptionOpts,
     ) => void;
```

Thoughts on this?